### PR TITLE
Moved ITestDataSource discovery to test discovery.

### PR DESCRIPTION
**Pull request by Haplois**
## Description
This PR moves discovery for `IDataSource` based data test methods to test discovery instead of execution. See the below examples for details.

### Work done:
1. `IDataSource` data tests (includes `DataRow` tests) are now discovered before test execution; and individually runnable.
2. Parent test results returned for data tests are not generated now. (See the new behavior of DataSourceAttribute in pictures.)

|       | IDataSource | DataSourceAttribute |
|----:|:------:|:------:|
| Current behavior | Current style with parent result (DataRow) | Current style with parent result (DataSource) |
| New behavior: | New implementation without a parent result (DataRow) | New implementation without a parent result (DataSource) |

### CODE REVIEW

File: src/Adapter/MSTest.CoreAdapter/Resources/Resource.resx [hunk 1]
```diff
@@ -320,4 +320,42 @@ Error: {1}</value>
   <data name="Execution_Test_Cancelled" xml:space="preserve">
     <value>Test '{0}' execution has been aborted.</value>
   </data>
+  <data name="CannotEnumarateDataSourceAttribute" xml:space="preserve">
```
Comment (peterwald) [inline src/Adapter/MSTest.CoreAdapter/Resources/Resource.resx]: `Enumerate` should be spelled with an 'e'.

File: src/Adapter/MSTest.CoreAdapter/Resources/Resource.resx [hunk 2]
```diff
@@ -320,4 +320,42 @@ Error: {1}</value>
...
+  <data name="FailedFetchAttributeCache" xml:space="preserve">
+    <value>Failed to get attribute cache. Ignoring attribute inheritance and falling into 'type defines Attribute model', so that we have some data.</value>
+  </data>
+  <data name="FailedToGetCustomAttribute" xml:space="preserve">
+    <value>Getting custom attributes for type {0} threw exception (will ignore and use the reflection way): {1}</value>
+    <comment>{0}: Attribute full type name.
+{1}: Exception description</comment>
+  </data>
+  <data name="OlderTFMVersionFound" xml:space="preserve">
+    <value>An older version of MSTestV2 package is loaded in asssembly, test discovery might fail to discover all data tests if they depend on `.runsettings` file.</value>
```
Comment (peterwald) [inline src/Adapter/MSTest.CoreAdapter/Resources/Resource.resx]: 2 's' in `assembly`

File: src/Adapter/MSTest.CoreAdapter/Helpers/DataSerializationHelper.cs [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+            return serializedData;
+        }
+
+        /// <summary>
+        /// Deserialzes the data serialzed by <see cref="Serialize(object[])" /> method.
+        /// </summary>
+        /// <param name="serializedData">Serialized data array to deserialize.</param>
+        /// <returns>Deserialized array.</returns>
+        public static object[] Deserialze(string[] serializedData)
```
Comment (peterwald) [inline src/Adapter/MSTest.CoreAdapter/Helpers/DataSerializationHelper.cs]: nit: `Deserialize()`

File: src/Adapter/MSTest.CoreAdapter/Helpers/DataSerializationHelper.cs [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+                if (serializedData[i] == null)
+                {
+                    data[i] = null;
+                    continue;
+                }
+
+                var serialzedDataBytes = Encoding.UTF8.GetBytes(serializedData[i]);
+                using (var memoryStream = new MemoryStream(serialzedDataBytes))
+                {
+                    data[i] = serializer.ReadObject(memoryStream);
```
Comment (peterwald) [inline src/Adapter/MSTest.CoreAdapter/Helpers/DataSerializationHelper.cs:74]: It would be nice to add an overload to the serializer to allow for ReadObject to operate on a `byte[]` or even better a `ReadOnlySpan<byte>`. That would eliminate the extra allocations here for the `MemoryStream`.

Same holds for the Serializer as well...adding a method to serialize directly to a `byte[]`

File: src/Adapter/MSTest.CoreAdapter/Helpers/ReflectHelper.cs [hunk 5]
```diff
@@ -113,7 +113,7 @@ public bool HasAttributeDerivedFrom(MemberInfo memberInfo, Type baseAttributeTyp
             Dictionary<string, object> attributes = this.GetAttributes(memberInfo, inherit);
             if (attributes == null)
             {
-                PlatformServiceProvider.Instance.AdapterTraceLogger.LogWarning("ReflectHelper.HasAttributeDerivedFrom: Failed to get attribute cache. Ignoring attribute inheritance and falling into 'type defines Attribute model', so that we have some data.");
+                PlatformServiceProvider.Instance.AdapterTraceLogger.LogWarning($"ReflectHelper.HasAttributeDerivedFrom: {Resource.FailedFetchAttributeCache}");
```
Comment (peterwald) [inline src/Adapter/MSTest.CoreAdapter/Helpers/ReflectHelper.cs]:
```csharp
$"{nameof(ReflectHelper)}.{nameof(HasAttributeDerivedFrom)}: {Resource.FailedFetchAttributeCache}"
```

would be more future-proof regarding renamings.

File: src/Adapter/MSTest.CoreAdapter/ObjectModel/UnitTestElement.cs [hunk 6]
```diff
@@ -188,16 +204,54 @@ internal TestCase ToTestCase()
...
+                testFullId += testCase.FullyQualifiedName;
+            }
+
+            if (this.TestMethod.DataType != DynamicDataType.None)
+            {
+                testFullId += testCase.DisplayName;
+            }
+
+            testCase.Id = EqtHash.GuidFromString(testFullId);
+
```
Comment (peterwald) [inline src/Adapter/MSTest.CoreAdapter/ObjectModel/UnitTestElement.cs:251]: Can we accumulate string values into the hash one by one, and grab the hash at the end? This code will allocate a new string each time a new value is appended. So a lot of string allocations for each id.

File: src/Adapter/MSTest.CoreAdapter/ObjectModel/UnitTestElement.cs [hunk 7]
```diff
@@ -188,16 +204,54 @@ internal TestCase ToTestCase()
...
         private string GetDisplayName()
         {
             if (string.IsNullOrWhiteSpace(this.DisplayName))
             {
-                return string.IsNullOrWhiteSpace(this.TestMethod.ManagedMethodName)
-                     ? this.TestMethod.Name
-                     : this.TestMethod.ManagedMethodName;
+                return this.TestMethod.Name;
+
+                // This causes compatibility problems with older runners.
```
Comment (peterwald) [inline src/Adapter/MSTest.CoreAdapter/ObjectModel/UnitTestElement.cs]: What is the compatibility issue? Is it because of unexpected characters? Or something different.

File: test/E2ETests/Smoke.E2E.Tests/CustomTestExecutionExtensibilityTests.cs [hunk 8]
```diff
@@ -40,8 +40,9 @@ public void ExecuteCustomTestExtensibilityWithTestDataTests()
                 "CustomTestMethod2 (B)",
                 "CustomTestMethod2 (B)");
 
+            // TODO :: ASK REVIEW :: SHOULD WE CONTINUE RETURNING A PARENT TEST FOR DATA BOUND TESTS?
             // Parent results should fail and thus failed count should be 7.
-            this.ValidateFailedTestsCount(7);
+            // this.ValidateFailedTestsCount(7);
```
Comment (peterwald) [inline test/E2ETests/Smoke.E2E.Tests/CustomTestExecutionExtensibilityTests.cs]: Do we not need this check any more?

File: test/E2ETests/Smoke.E2E.Tests/CustomTestExecutionExtensibilityTests.cs [hunk 9]
```diff
@@ -40,8 +40,9 @@ public void ExecuteCustomTestExtensibilityWithTestDataTests()
                 "CustomTestMethod2 (B)",
                 "CustomTestMethod2 (B)");
 
+            // TODO :: ASK REVIEW :: SHOULD WE CONTINUE RETURNING A PARENT TEST FOR DATA BOUND TESTS?
```
Comment (peterwald) [inline test/E2ETests/Smoke.E2E.Tests/CustomTestExecutionExtensibilityTests.cs]: I don't think we should continue to return parent test, as there really is no such concept elsewhere. It was primarily used to support the summary feature which is no longer helpful.

**Review by peterwald — APPROVED** [review]
Looks like `test/E2ETests/DiscoveryAndExecutionTests/Utilities/CLITestBase.discovery.cs` is empty?

Mostly LGTM with comments noted.

File: src/Adapter/MSTest.CoreAdapter/ObjectModel/UnitTestElement.cs (hunk 7 above)
Comment (Haplois) [reply src/Adapter/MSTest.CoreAdapter/ObjectModel/UnitTestElement.cs]: Yes, basically because unexpected and unescaped char in name https://github.com/microsoft/testfx/issues/789

File: test/E2ETests/Smoke.E2E.Tests/CustomTestExecutionExtensibilityTests.cs (hunk 8 above)
Comment (Haplois) [reply test/E2ETests/Smoke.E2E.Tests/CustomTestExecutionExtensibilityTests.cs]: We're not returning a parent test anymore.

**Comment by Haplois** [comment]
> @peterwald: 
> Looks like `test/E2ETests/DiscoveryAndExecutionTests/Utilities/CLITestBase.discovery.cs` is empty?

It actually is not; I don't know why it's listed as a binary file...

File: scripts/Build.ps1 [hunk 10]
```diff
@@ -48,6 +48,9 @@ Param(
 
   [Alias("tpv")]
   [string] $TestPlatformVersion = $null,
+  
+  [Alias("np")]
+  [Switch] $DisallowPrereleaseMSBuild,
```
Comment (shyamnamboodiripad) [inline scripts/Build.ps1]: This seems to be duplicated below. Was this intentional?

File: src/Adapter/MSTest.CoreAdapter/Discovery/AssemblyEnumerator.cs [hunk 11]
```diff
@@ -40,6 +53,11 @@ public AssemblyEnumerator(MSTestSettings settings)
             MSTestSettings.PopulateSettings(settings);
         }
 
+        /// <summary>
+        /// Sets run settings to use for current discovery session.
+        /// </summary>
+        public string RunSettingsXml { private get; set; }
```
Comment (shyamnamboodiripad) [inline src/Adapter/MSTest.CoreAdapter/Discovery/AssemblyEnumerator.cs]: nit: Might as well make the getter also public. I find set-only properties a bit weird :) although this is not in public API so may be ok...

File: src/Adapter/MSTest.CoreAdapter/Discovery/AssemblyEnumerator.cs [hunk 12]
```diff
@@ -55,14 +73,15 @@ public override object InitializeLifetimeService()
...
-        /// <returns> A collection of Test Elements. </returns>
-        [SuppressMessage("Microsoft.Design", "CA1031:DoNotCatchGeneralExceptionTypes", Justification = "Catching a generic exception since it is a requirement to not abort discovery in case of any errors.")]
+        /// <param name="assemblyFileName">The assembly file name.</param>
+        /// <param name="warnings">Contains warnings if any, that need to be passed back to the caller.</param>
+        /// <returns>A collection of Test Elements.</returns>
         internal ICollection<UnitTestElement> EnumerateAssembly(string assemblyFileName, out ICollection<string> warnings)
         {
             Debug.Assert(!string.IsNullOrWhiteSpace(assemblyFileName), "Invalid assembly file name.");
 
+            string runSettingsXml = this.RunSettingsXml;
```
Comment (shyamnamboodiripad) [inline src/Adapter/MSTest.CoreAdapter/Discovery/AssemblyEnumerator.cs]: nit: var

Also this local can probably be deleted i.e. the callsite below can probably access the field directly.

File: src/Adapter/MSTest.CoreAdapter/Discovery/AssemblyEnumerator.cs [hunk 13]
```diff
@@ -6,21 +6,34 @@ namespace Microsoft.VisualStudio.TestPlatform.MSTest.TestAdapter.Discovery
...
+
     /// <summary>
     /// Enumerates through all types in the assembly in search of valid test methods.
     /// </summary>
     internal class AssemblyEnumerator : MarshalByRefObject
     {
+        /// <summary>
+        /// Helper for reflection API's.
+        /// </summary>
+        private static readonly ReflectHelper ReflectHelper = new ReflectHelper();
```
Comment (shyamnamboodiripad) [inline src/Adapter/MSTest.CoreAdapter/Discovery/AssemblyEnumerator.cs]: Consider: Perhaps you could define a static instance in the type itself that all callers can then use i.e. `ReflectHelper.Instance`

File: src/Adapter/MSTest.CoreAdapter/Discovery/AssemblyEnumeratorWrapper.cs [hunk 14]
```diff
@@ -22,8 +21,7 @@ internal class AssemblyEnumeratorWrapper
         /// <summary>
         /// Assembly name for UTF
         /// </summary>
-        private static readonly AssemblyName UnitTestFrameworkAssemblyName =
-            typeof(TestMethodAttribute).GetTypeInfo().Assembly.GetName();
+        private static readonly AssemblyName UnitTestFrameworkAssemblyName = typeof(TestMethodAttribute).GetTypeInfo().Assembly.GetName();
```
Comment (shyamnamboodiripad) [inline src/Adapter/MSTest.CoreAdapter/Discovery/AssemblyEnumeratorWrapper.cs]: nit: I would recommend moving refactoring / formatting / code hygiene changes that are unrelated to the primary change to a separate PR (especially when there are many such refactoring changes). Otherwise, it can be difficult for reviewers to figure out and focus on exactly what is changing 😄

File: src/Adapter/MSTest.CoreAdapter/Discovery/AssemblyEnumeratorWrapper.cs [hunk 15]
```diff
@@ -71,33 +59,18 @@ internal class AssemblyEnumeratorWrapper
...
-                    fullFilePath,
-                    ex);
-
-                var message = string.Format(
-                    CultureInfo.CurrentCulture,
-                    Resource.TestAssembly_AssemblyDiscoveryFailure,
-                    fullFilePath,
-                    ex.Message);
+                var message = string.Format(CultureInfo.CurrentCulture, Resource.TestAssembly_AssemblyDiscoveryFailure, fullFilePath, ex.Message);
+                PlatformServiceProvider.Instance.AdapterTraceLogger.LogWarning($"{nameof(AssemblyEnumeratorWrapper)}.{nameof(this.GetTests)}: {Resource.TestAssembly_AssemblyDiscoveryFailure}", fullFilePath, ex);
```
Comment (shyamnamboodiripad) [inline src/Adapter/MSTest.CoreAdapter/Discovery/AssemblyEnumeratorWrapper.cs]: Is it by design that the trace messages are now localized? I am assuming the trace messages were mainly intended for our own debugging (as opposed to for a user trying to debug problems on their own). If yes, then localizing the trace logs may make it more difficult to analyze logs supplied by users...

Just checking to make sure this was an intentional change that has been considered by the team.

File: src/Adapter/MSTest.CoreAdapter/Discovery/AssemblyEnumeratorWrapper.cs [hunk 16]
```diff
@@ -141,8 +106,17 @@ private ICollection<UnitTestElement> GetTestsInIsolation(string fullFilePath, IR
             using (var isolationHost = PlatformServiceProvider.Instance.CreateTestSourceHost(fullFilePath, runSettings, frameworkHandle: null))
             {
                 // Create an instance of a type defined in adapter so that adapter gets loaded in the child app domain
-                var assemblyEnumerator = isolationHost.CreateInstanceForType(
-                    typeof(AssemblyEnumerator), new object[] { MSTestSettings.CurrentSettings }) as AssemblyEnumerator;
+                var assemblyEnumerator = isolationHost.CreateInstanceForType(typeof(AssemblyEnumerator), new object[] { MSTestSettings.CurrentSettings }) as AssemblyEnumerator;
+
+                // This might not be supported if an older version of "PlatformServices
```
Comment (shyamnamboodiripad) [inline src/Adapter/MSTest.CoreAdapter/Discovery/AssemblyEnumeratorWrapper.cs]: Comment seems incomplete. Was some text deleted by accident?

File: src/Adapter/MSTest.CoreAdapter/Discovery/TypeEnumerator.cs [hunk 17]
```diff
@@ -199,6 +198,8 @@ internal UnitTestElement GetTestFromMethod(MethodInfo method, bool isDeclaredInT
                 testElement.WorkItemIds = workItemAttributes.Select(x => x.Id.ToString()).ToArray();
             }
 
+            testElement.Ignored = this.reflectHelper.IsAttributeDefined(method, typeof(IgnoreAttribute), false);
```
Comment (shyamnamboodiripad) [inline src/Adapter/MSTest.CoreAdapter/Discovery/TypeEnumerator.cs:201]: Just curious. Was it a bug that we were not populating ignore info before? Or did this code exist elsewhere before?

File: src/Adapter/MSTest.CoreAdapter/Discovery/AssemblyEnumerator.cs [hunk 18]
```diff
@@ -220,11 +196,191 @@ internal string GetLoadExceptionDetails(ReflectionTypeLoadException ex)
...
+                            tests.Add(test);
+                            continue;
+                        }
+
+                        if (this.DynamicDataAttached(sourceLevelParameters, assembly, test, tests))
+                        {
+                            continue;
+                        }
+
+                        tests.Add(test);
```
Comment (shyamnamboodiripad) [inline src/Adapter/MSTest.CoreAdapter/Discovery/AssemblyEnumerator.cs:234]: The above logic is slightly confusing. Perhaps adding comments would help.

I am not sure I understand why its important to check whether test is ignored in this case - in other words, why are dynamic data tests only added if they are ignored (and skipped otherwise)?

Why can't we do the following instead? In other words, can we skip all dynamic tests and add add all remaining tests (regardless of whether they are ignored).

```
foreach (var test in unitTestCases)
{
    if (this.DynamicDataAttached(sourceLevelParameters, assembly, test, tests))
    {
        continue;
    }

    tests.Add(test);
}
```

File: src/Adapter/MSTest.CoreAdapter/Discovery/AssemblyEnumerator.cs [hunk 19]
```diff
@@ -220,11 +196,191 @@ internal string GetLoadExceptionDetails(ReflectionTypeLoadException ex)
...
+                    }
+                }
+            }
+            catch (Exception exception)
+            {
+                // If we fail to discover type from a class, then don't abort the discovery
+                // Move to the next type.
+                string message = string.Format(CultureInfo.CurrentCulture, Resource.CouldNotInspectTypeDuringDiscovery, typeFullName, assemblyFileName, exception.Message);
+                PlatformServiceProvider.Instance.AdapterTraceLogger.LogInfo($"AssemblyEnumerator: {message}");
+                warningMessages.Add(message);
```
Comment (shyamnamboodiripad) [inline src/Adapter/MSTest.CoreAdapter/Discovery/AssemblyEnumerator.cs:244]: Will this warning be relayed to the Test Explorer when client is VS?

File: src/Adapter/MSTest.CoreAdapter/Discovery/AssemblyEnumerator.cs [hunk 20]
```diff
@@ -220,11 +196,191 @@ internal string GetLoadExceptionDetails(ReflectionTypeLoadException ex)
...
+            }
+
+            return tests;
+        }
+
+        private bool DynamicDataAttached(IDictionary<string, object> sourceLevelParameters, Assembly assembly, UnitTestElement test, List<UnitTestElement> tests)
+        {
+            if (test.TestMethod.HasManagedMethodAndTypeProperties == false)
```
Comment (shyamnamboodiripad) [inline src/Adapter/MSTest.CoreAdapter/Discovery/AssemblyEnumerator.cs]: When is this expected to be false? Aren't all test methods supposed to have this now? Perhaps a comment would help...

Comment (shyamnamboodiripad) [inline src/Adapter/MSTest.CoreAdapter/Discovery/AssemblyEnumerator.cs]: minor: `if (!test.TestMethod.HasManagedMethodAndTypeProperties)` (assuming this is a simple bool)

File: src/Adapter/MSTest.CoreAdapter/Discovery/AssemblyEnumerator.cs [hunk 21]
```diff
@@ -220,11 +196,191 @@ internal string GetLoadExceptionDetails(ReflectionTypeLoadException ex)
...
+                var testContext = PlatformServiceProvider.Instance.GetTestContext(testMethod, writer, sourceLevelParameters);
+                var testMethodInfo = this.typeCache.GetTestMethodInfo(testMethod, testContext, MSTestSettings.CurrentSettings.CaptureDebugTraces);
+                if (testMethodInfo == null)
+                {
+                    return false;
+                }
+
+                return false /* DataSourceAttribute discovery is disabled for now, since we cannot serialize DataRow values.
+                    || this.TryProcessDataSource(test, testMethodInfo, testContext, tests) */
+                    || this.TryProcessTestDataSourceTests(test, testMethodInfo, tests);
```
Comment (shyamnamboodiripad) [inline src/Adapter/MSTest.CoreAdapter/Discovery/AssemblyEnumerator.cs]: Can we omit the false here?

```
return /* DataSourceAttribute discovery is disabled for now, since we cannot serialize DataRow values.
    this.TryProcessDataSource(test, testMethodInfo, testContext, tests) || */
    this.TryProcessTestDataSourceTests(test, testMethodInfo, tests);
```

File: src/Adapter/MSTest.CoreAdapter/Discovery/AssemblyEnumerator.cs [hunk 22]
```diff
@@ -220,11 +196,191 @@ internal string GetLoadExceptionDetails(ReflectionTypeLoadException ex)
...
+            }
+        }
+
+        private bool TryProcessDataSource(UnitTestElement test, TestMethodInfo testMethodInfo, ITestContext testContext, List<UnitTestElement> tests)
+        {
+            UTF.DataSourceAttribute[] dataSourceAttributes = ReflectHelper.GetAttributes<UTF.DataSourceAttribute>(testMethodInfo.MethodInfo, false);
```
Comment (shyamnamboodiripad) [inline src/Adapter/MSTest.CoreAdapter/Discovery/AssemblyEnumerator.cs]: consider: `var` (also other occurrences below)

File: src/Adapter/MSTest.CoreAdapter/Discovery/AssemblyEnumerator.cs [hunk 23]
```diff
@@ -220,11 +196,191 @@ internal string GetLoadExceptionDetails(ReflectionTypeLoadException ex)
...
+                    return this.ProcessDataSourceTests(test, testMethodInfo, testContext, tests);
+                }
+                catch (Exception ex)
+                {
+                    var message = string.Format(CultureInfo.CurrentCulture, Resource.CannotEnumerateDataSourceAttribute, test.TestMethod.ManagedTypeName, test.TestMethod.ManagedMethodName, ex);
+                    PlatformServiceProvider.Instance.AdapterTraceLogger.LogInfo($"DynamicDataEnumarator: {message}");
+                    return false;
+                }
+            }
+            else if (dataSourceAttributes != null && dataSourceAttributes.Length > 1)
```
Comment (shyamnamboodiripad) [inline src/Adapter/MSTest.CoreAdapter/Discovery/AssemblyEnumerator.cs]: consider: Add early outs

```
if (dataSourceAttributes == null || dataSourceAttributes.Length == 0)
{
    return false;
}

if (dataSourceAttributes.Length > 1)
{
    ...
    throw new InvalidOperationException(...);
}

// dataSourceAttributes.Length == 1
try
{
    return this.ProcessDataSourceTests(test, testMethodInfo, testContext, tests);
}
catch
{
    ...
    return false;
}
```

File: src/Adapter/MSTest.CoreAdapter/Discovery/AssemblyEnumerator.cs [hunk 24]
```diff
@@ -220,11 +196,191 @@ internal string GetLoadExceptionDetails(ReflectionTypeLoadException ex)
...
+            }
+
+            try
+            {
+                int rowIndex = 0;
+
+                foreach (var dataRow in dataRows)
+                {
+                    // TODO: Test serialization
+                    rowIndex++;
```
Comment (shyamnamboodiripad) [inline src/Adapter/MSTest.CoreAdapter/Discovery/AssemblyEnumerator.cs:317]: consider: Use a `for` loop in place of `foreach`

File: src/Adapter/MSTest.CoreAdapter/Discovery/AssemblyEnumerator.cs [hunk 25]
```diff
@@ -220,11 +196,191 @@ internal string GetLoadExceptionDetails(ReflectionTypeLoadException ex)
...
+            }
+
+            return false;
+        }
+
+        private bool TryProcessTestDataSourceTests(UnitTestElement test, TestMethodInfo testMethodInfo, List<UnitTestElement> tests)
+        {
+            var methodInfo = testMethodInfo.MethodInfo;
+
+            UTF.ITestDataSource[] testDataSources = ReflectHelper.GetAttributes<Attribute>(methodInfo, false)?.Where(a => a is UTF.ITestDataSource).OfType<UTF.ITestDataSource>().ToArray();
```
Comment (shyamnamboodiripad) [inline src/Adapter/MSTest.CoreAdapter/Discovery/AssemblyEnumerator.cs]: Looks like testDataSources can be null. Do we need an early return for that?

File: src/Adapter/MSTest.CoreAdapter/Discovery/AssemblyEnumerator.cs [hunk 26]
```diff
@@ -220,11 +196,191 @@ internal string GetLoadExceptionDetails(ReflectionTypeLoadException ex)
...
+                    var discoveredTest = test.Clone();
+                    discoveredTest.DisplayName = displayName;
+                    discoveredTest.TestMethod.DataType = DynamicDataType.DataSourceAttribute;
+                    discoveredTest.TestMethod.Data = new[] { (object)rowIndex };
+                    tests.Add(discoveredTest);
+                }
+
+                return true;
+            }
+            catch
```
Comment (shyamnamboodiripad) [inline src/Adapter/MSTest.CoreAdapter/Discovery/AssemblyEnumerator.cs]: Should this exception be logged (similar to other methods above and below)?

File: test/E2ETests/Smoke.E2E.Tests/DataRowTests.cs [hunk 27]
```diff
@@ -26,7 +26,7 @@ public void ExecuteOnlyDerivedClassDataRowsWhenBothBaseAndDerviedClassHasDataRow
             // 4 tests of BaseClass.DataRowTestMethod - 3 data row results and 1 parent result
             // 3 tests of DerivedClass.DataRowTestMethod - 2 data row results and 1 parent result
             // Total 7 tests - Making sure that DerivedClass doesn't run BaseClass tests
-            this.ValidatePassedTestsCount(7);
+            this.ValidatePassedTestsCount(7 - 2);
```
Comment (shyamnamboodiripad) [inline test/E2ETests/Smoke.E2E.Tests/DataRowTests.cs]: Any reason to not just say 5 instead of 7 - 2?

(Same question for below changes)

File: src/Adapter/MSTest.CoreAdapter/Helpers/ReflectHelper.cs [hunk 28]
```diff
@@ -200,6 +200,20 @@ public override object InitializeLifetimeService()
...
+        {
+            Attribute[] attributeArray = GetCustomAttributes(methodBase, typeof(T), inherit);
+
+            var attributes = attributeArray as T[];
+            if (attributes != null)
+            {
+                return attributes;
+            }
+
+            return attributeArray?.Select(a => a as T)?.Where(a => a != null)?.ToArray();
```
Comment (shyamnamboodiripad) [inline src/Adapter/MSTest.CoreAdapter/Helpers/ReflectHelper.cs]: This can be simplified as `return attributeArray?.Where(a => a is T).ToArray()`

File: src/Adapter/MSTest.CoreAdapter/Helpers/ReflectHelper.cs [hunk 29]
```diff
@@ -200,6 +200,20 @@ public override object InitializeLifetimeService()
             return null;
         }
 
+        internal static T[] GetAttributes<T>(MethodBase methodBase, bool inherit)
+            where T : Attribute
+        {
+            Attribute[] attributeArray = GetCustomAttributes(methodBase, typeof(T), inherit);
+
+            var attributes = attributeArray as T[];
```
Comment (shyamnamboodiripad) [inline src/Adapter/MSTest.CoreAdapter/Helpers/ReflectHelper.cs]: Looks like this will never succeed since GetCustomAttributes below always returns `attributesArray.OfType<Attribute>().ToArray();`. Consider removing this block to avoid confusion.

File: src/Adapter/MSTest.CoreAdapter/Helpers/ReflectHelper.cs (hunk 28 above)
Comment (shyamnamboodiripad) [inline src/Adapter/MSTest.CoreAdapter/Helpers/ReflectHelper.cs]: Would it make sense to add an early return for `attributeArrray == null` above and omit the `?.` operators on this line?

**Review by shyamnamboodiripad — COMMENTED** [review]
I haven't finished reviewing all the changes but wanted to post the comments so far. Will continue reviewing the remaining code tomorrow.

File: src/Adapter/MSTest.CoreAdapter/Discovery/TypeEnumerator.cs (hunk 17 above)
Comment (Haplois) [reply src/Adapter/MSTest.CoreAdapter/Discovery/TypeEnumerator.cs:201]: No, this was set during execution.

File: src/Adapter/MSTest.CoreAdapter/Discovery/AssemblyEnumerator.cs (hunk 18 above)
Comment (Haplois) [reply src/Adapter/MSTest.CoreAdapter/Discovery/AssemblyEnumerator.cs:234]: So, should we discover all tests, even if they are ignored or in an ignored class?

File: scripts/Build.ps1 (hunk 10 above)
Comment (Haplois) [reply scripts/Build.ps1]: No, it was a merge error, fixed.

File: src/Adapter/MSTest.CoreAdapter/Discovery/AssemblyEnumeratorWrapper.cs (hunk 14 above)
Comment (Haplois) [reply src/Adapter/MSTest.CoreAdapter/Discovery/AssemblyEnumeratorWrapper.cs]: Thanks for pointing my laziness out :D I moved all those into a separate PR, though it's too late for your review now :)

**Comment by AbhitejJohn** [comment]
Is this going to be configurable through runsettings? For instance, Can a user choose not to discover tests from a large excel file if they do not want to unless they run it? That would also be useful given this changes the public extensibility layer.

**Comment by shyamnamboodiripad** [comment]
> Is this going to be configurable through runsettings? For instance, Can a user choose not to discover tests from a large excel file if they do not want to unless they run it? That would also be useful given this changes the public extensibility layer.

That's a good point and I understand why this could be useful. However, if the test is not split at discovery time, this means that to run a single dynamic test one would have to always re-discover and run all of them. This could be equally annoying I think. A setting would also add concept count + difference in behavior when the setting is on v/s off could cause confusion as one moves between different repos.

If we do add such a setting, we should probably limit it to be applicable for dynamic tests only. Tests with data points defined in source should be expanded regardless.

I also think a better alternative for the issue you raise would be to fully decouple test discovery from build in VS. If there were a way to do that (and trigger discovery explicitly via some sort of refresh button or via the run all tests button), one would avoid paying any discovery cost until one is ready to run tests, while at the same time retaining the ability to run individual dynamic tests.

@AbhitejJohn @Haplois Do you know if other frameworks (xUnit / NUnit) allow to configure the splitting behavior?

File: src/Adapter/MSTest.CoreAdapter/Discovery/AssemblyEnumerator.cs (hunk 18 above)
Comment (shyamnamboodiripad) [reply src/Adapter/MSTest.CoreAdapter/Discovery/AssemblyEnumerator.cs:234]: I think we should discover all tests. Ignored tests should be discovered, but should return the "skipped" status upon execution (and should be displayed with a yellow skipped icon in VS). This would match the behavior for other frameworks.

If we skip discovery for ignored tests they would not be displayed in VS. This means it would be impossible to un-skip and execute them again without running a build to make them appear first after un-skipping.

One may also want to know how many tests in a given solution are currently ignored by looking at total count of ignored tests in VS (and fix them up / un-skip by navigating to them one-by-one). This would again not be impossible if the skipped tests are not discovered in the first place.

Some frameworks also have a conditional skip attribute (e.g. skip if OS=Linux) where the condition is evaluated at run time. I am not sure if mstest has such capability - but this would be another reason to only skip at run time.

File: src/Adapter/MSTest.CoreAdapter/Discovery/TypeEnumerator.cs (hunk 17 above)
Comment (shyamnamboodiripad) [reply src/Adapter/MSTest.CoreAdapter/Discovery/TypeEnumerator.cs:201]: Ah thanks please see the other comment around ignores.

File: src/Adapter/MSTest.CoreAdapter/Discovery/AssemblyEnumerator.cs (hunk 20 above)
Comment (Haplois) [reply src/Adapter/MSTest.CoreAdapter/Discovery/AssemblyEnumerator.cs]: It should always be `true`, but any part of the chain is obsolete; it might not contain those. Since we depend on those properties, if they don't exist, we bail out early—added comment.

File: src/Adapter/MSTest.CoreAdapter/Discovery/AssemblyEnumerator.cs (hunk 24 above)
Comment (Haplois) [reply src/Adapter/MSTest.CoreAdapter/Discovery/AssemblyEnumerator.cs:317]: `dataRows` is `IEnumerable<object>`, `foreach` is more convinient here.

File: src/Adapter/MSTest.CoreAdapter/Discovery/AssemblyEnumerator.cs (hunk 26 above)
Comment (Haplois) [reply src/Adapter/MSTest.CoreAdapter/Discovery/AssemblyEnumerator.cs]: Converted to `finally` since we already `catch` the error and log it outer levels.

File: src/Adapter/MSTest.CoreAdapter/Helpers/DataSerializationHelper.cs (hunk 4 above)
Comment (Haplois) [reply src/Adapter/MSTest.CoreAdapter/Helpers/DataSerializationHelper.cs:74]: `DataContractJsonSerializer` doesn't support `Span`.

File: src/Adapter/MSTest.CoreAdapter/Helpers/ReflectHelper.cs (hunk 29 above)
Comment (Haplois) [reply src/Adapter/MSTest.CoreAdapter/Helpers/ReflectHelper.cs]: In `~/src/Adapter/PlatformServices.Desktop/Utilities/DesktopReflectionUtility.cs` it might return a `T[]`, and that might be the most common path for this method.
